apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabTheorycraft/optimizeTc.ts

```diff
@@ -99,70 +100,103 @@ export function optimizeTc(
+  //   distributed = assignableMaxTot
+  //   maxBuffer = Object.fromEntries(
+  //     subsArr.map((x) => [x, substatValue(x, maxSubstats[x])])
+  //   )
+  //   if (shouldShowDevComponents)
+  //     console.log({ maxBuffer, subsArr, maxSubstats, distributed })
+  // } else {
+  let max = -Infinity
+  const buffer = {} //Object.fromEntries([...subs].map((x) => [x, 0]))
+  const bufferInt = {} // Object.fromEntries([...subs].map((x) => [x, 0]))
+  const existingSubs = objMap(
+    charTC.artifact.substats.stats,
+    (v, k) => toDecimal(v, k) + minTotalRolls[k]
+  )
+  const allKeys = [...allSubstatKeys, 'other']
+  const permute = (toAssign: number, [x, ...xs]: string[]) => {
+    if (xs.length === 0) {
+      if (toAssign > maxSubsAssignable[x]) return
+
+      if (x !== 'other') buffer[x] = substatValue(x, toAssign)
+      bufferInt[x] = toAssign
+
+      const allRolls = allKeys.map((k) => [
+        k,
+        (existingSubs[k] ?? 0) + (bufferInt[k] ?? 0),
+      ]) as Array<[SubstatKey | 'other', number]>
+      if (!isValid(allRolls)) return
+      const [result] = compute([
+        { values: existingSubs },
+        { values: buffer },
+      ] as const)
+      if (result > max) {
+        console.log(result)
+        max = result
+        maxBuffer = structuredClone(buffer)
+        maxBufferInt = structuredClone(bufferInt)
       }
+      return
     }
-    permute(distributedSubstats, subsArr)
-    if (shouldShowDevComponents) {
-      console.log(`Took ${performance.now() - startTime} ms`)
-      console.log({
-        maxBuffer,
-        maxBufferInt: objMap(
-          maxBuffer,
-          (v, k) =>
-            v / getSubstatValue(k as SubstatKey, rarity, substatsType, false)
-        ),
-        subsArr,
-      })
+    for (let i = 0; i <= Math.min(maxSubsAssignable[x], toAssign); i++) {
+      // TODO: Making sure that i + \sum { maxSubstats[xs] } >= distributedSubstats in each recursion will reduce unnecessary recursion considerably for large problems. It will also tighten the possibilities for the leaf recursion, so you don't need so many checkings.
+      // https://github.com/frzyc/genshin-optimizer/pull/781#discussion_r1138083742
+      buffer[x] = substatValue(x, i)
+      bufferInt[x] = i
+      permute(toAssign - i, xs)
     }
   }
+  permute(distributedSubstats, [...scalesWith, 'other'])
+  if (shouldShowDevComponents) {
+    console.log(`Took ${performance.now() - startTime} ms`)
+    console.log({
+      maxBuffer,
+      maxBufferInt,
+      scalesWith,
+    })
+  }
+  distributed += maxBufferInt.other
+  // }
   return {
     maxBuffer,
     distributed,
-    scalesWith: subsArr.filter((x) =>
-      (allSubstatKeys as readonly string[]).includes(x)
-    ) as SubstatKey[],
+    scalesWith,
+  }
+}
+
+// assumes the rolls respect mainstat max rolls, and max total rolls, and the distribution will be above the min #rolls for rarity
+function isValid(subsRolls: Array<[SubstatKey | 'other', number]>) {
```

Comment:
This check the feasibility of occupying all 20 substat slots, by stacking as much scaling subs in minimal slots as possible, and check if there is enough "off-scaling" subs("other") to occupy the remaining spots